### DISCUSSION TITLE
Mainline pygls in vscode.dev!

### DISCUSSION THREAD

**Discussion by alcarney** [openlawlibrary/pygls#510] [Show and tell]
![image](https://github.com/user-attachments/assets/decf9a6c-a08a-4e5c-a063-a8438362da39)
*Screenshot shows running our example `semantic_tokens.py` server running in the browser*

With the recent changes to the pygls `main` branch (i.e. the upcoming v2 release) I'm pleased to report that it is now possible to run your language server fully in the browser in https://vscode.dev! :smile: 

Rather than using Pyodide as the WASM runtime, this is using [VSCode's WASI host](https://code.visualstudio.com/blogs/2023/06/05/vscode-wasm-wasi), as such your server will have to align with the current limitations of the WASI runtime

- No async-await support
- No threads

See [this repository](https://github.com/alcarney/vscode-wasi-pygls/) for details on how to make it work.

**Comment by tombh** [comment]
Wow, this is amazing.

That link you have in the screenshot is this right? https://vscode.dev/github/openlawlibrary/pygls I assume it's expected that merely going to that URL isn't enough to set everything up? I get this error when visiting it: 
![image](https://github.com/user-attachments/assets/1284ea34-923d-49db-b6c4-4dbee697165f)

So are you planning to write up some steps at https://github.com/alcarney/vscode-wasi-pygls?

**Reply by alcarney** [reply]
> I get this error when visiting it:

Are you using firefox? I saw the same, had to switch to chromium to get VSCode to load properly :/ 

> So are you planning to write up some steps at https://github.com/alcarney/vscode-wasi-pygls?

Yes, I suppose that would help! :sweat_smile: 

The short version is though
- Open https://vscode.dev/github/openlawlibrary/pygls, if vscode.dev is working properly, it should ask you to sign into GitHub before loading the repo.
- Install the `swyddfa.pygls-wasi-playground` extension
- Tweak `.vscode/settings.json` to match what you see in the screenshot.
- You'll likely have encountered an error similar to the one you see here https://github.com/alcarney/vscode-wasi-pygls/issues/5 you'll have to run the `Developer: Reload Window` command to work around it for now
- Wait for the server to boot up (will take 2-3min), it might be worth having the network tab in the devtools open to get an idea for how progress is going

**Reply by tombh** [reply]
Nice, I made some progress, but seems to get stuck on 404s for `.pyc` files? 
![image](https://github.com/user-attachments/assets/2b94b8f2-d0b9-4528-9606-b0d02a7a173e)

**Reply by alcarney** [reply]
I saw those also, but they didn't seem to cause an issue for me - I assume that's just Python checking for pre-compiled `*.pyc` files as it would normally.

Since the code in your screenshot is not highlighted, I assume the server didn't start? Is there any output in the `pygls-client` or `pygls-server` output channels?

**Reply by alcarney** [reply]
Just published a new version of the extension with some improvements, it should now be possible to use the `settings.json` file (almost) as-is. The only tweak that is still needed is to remove `"scheme": "file"` from the `pygls.client.documentSelector` option
```diff
 {
     "[plaintext]": {
         // Uncomment to enable `textDocument/onTypeFormatting` requests
         // "editor.formatOnType": true
     },
     // Uncomment to override Python interpreter used.
     // "pygls.server.pythonPath": "/path/to/python",
     "pygls.server.debug": false,
     // "pygls.server.debugHost": "localhost",
     // "pygls.server.debugPort": 5678,
     "pygls.server.launchScript": "code_actions.py", // This is relative to `pygls.server.cwd`
     "pygls.server.cwd": "${workspaceFolder}/examples/servers",
     "pygls.trace.server": "off",
     "pygls.client.documentSelector": [
         {
-            "scheme": "file",
             "language": "plaintext"
         }
     ],
     "python.testing.pytestArgs": [
         "."
     ],
     "python.testing.unittestEnabled": false,
     "python.testing.pytestEnabled": true,
     // "pygls.jsonServer.exampleConfiguration": "some value here",
 }

```

**Comment by tombh** [comment]
Nice! I got it to work this time! Evaluated `1 + 1 =` .... `2` 😏

All in the browser, wow.
